text/0045-apply-keyword.md

```diff
@@ -0,0 +1,119 @@
+
+## Timeline
+
+- Started: 2024-01-25
+
+## Summary
+
+Add a keyword that allows denoted policy statements to have no impact on the
+authorization decision, yet be logged in the diagnostics for later annotation
+aggregation.
+
+## Basic example
+
+```cedar
+@annotation("value")
+apply (
+    principal,
+    action,
+    resource
+);
+```
+
+## Motivation
+
+I want to be able to create policy statements that will add annotations while
+not affecting the authorization outcome.  Here are some examples:
+
+1. I want to be able to require MFA anytime someone accesses a certain table.
+However, I want to be able to define permission for this table elsewhere.
+
+2. I want to be able to send out an email everytime something is denied.
+However, I want to be able to define what things are forbidden elsewhere.
+
+3. I want to be able to send out an email regardless of if a specific event
+is allowed or denied.
+
+## Detailed design
+
+There are three cases that are added to the language syntax:
+
+1. Apply on permit
+
+```cedar
+@mfa("true")
+apply on permit (
+    principal,
+    action == Postgres::Action::"select",
+    resource == Postgres::Table::"example.com:5432/db"
+);
+```
+
+2. Apply on forbid
+
+```cedar
+@email("alert@example.com")
+apply on forbid (
+    principal,
+    action,
+    resource
+);
+```
+
+3. Always apply
+
+```cedar
+@email("alert@example.com")
+apply (
+    principal,
+    action,
+    resource in Postgres::Table::"example.com:5432/db"
+);
+```
+
+The apply keyword will cause all matching policy statements to be added to the
+diagnostics, while not impacting the authorization decision in any way.
+
+### Changes required to parser/AST
+
+- Update grammer to support `apply on permit` `apply on forbid` `apply` in
+  addition to the existing `forbid` and `permit`
+
+### Changes required to evaluator
+
+- Policies that are categorized as `apply` would not impact the decision, while matches
+would be added to the diagnostic.
+
+## Drawbacks
+
+Why should we *not* do this? Please consider:
+
+- this does add the `apply` and `on` keyword to the grammar, thus increasing the
+complexity of the Cedar language
+- this may have some impact on the ability to short-circuit policy decisions,
+since all apply statements need to be checked to be added to the diagnostics
+
+## Alternatives
+
+I have considered using an annotation `@apply("after")` as a way of implementing
+this.  This works for the `apply on permit` case, although it changes the
+outcome of the authorization event.  By taking into consideration the
+annotation, I am able to reverse the decision to `deny` if all matching policies
+are annotated with `@apply("after")`.
+
+This alternative does not work with the `forbid` case because any matching
+forbid statements will cause the entire request to be denied, regardless of if
+the principal should be permitted.
+
+This alternative does not support the general case of applying regardless of
+permit / forbid.
```

Comment:
Hi @philhassey! Would this workaround work if the `is_authorized` function returned more information? Currently the `Response` will store the list of satisfied `permit` policies (for an Allow decision) or the list of satisfied `forbid` polcies (for a Deny decision). What if the `Response` instead specified all satisfied policies, regardless of decision?

Reply:
Yes this could work, I would just throw away the given decision from `is_authorized` and come up with my own decision based on the annotations.